src/WopiValidator.Core/Validators/JsonContentValidator.cs

```diff
@@ -314,15 +316,32 @@ public override bool Validate(JToken actualValue, Dictionary<string, string> sav
 					return false;
 
 				errorMessage = "";
-				if (String.IsNullOrWhiteSpace(_endsWithValue))
-					return true;
-
 				string typedActualValue = actualValue.Value<string>();
 				string formattedActualValue = FormatValue(typedActualValue);
 
-				if (!formattedActualValue.EndsWith(_endsWithValue))
+				if (!String.IsNullOrWhiteSpace(_endsWithValue))
+				{
+					if (!formattedActualValue.EndsWith(_endsWithValue))
+					{
+						errorMessage = string.Format("Expected to end with: '{0}', Actual: '{1}'", _endsWithValue, formattedActualValue);
+						return false;
+					}
+				}
+
+				string setting = null;
+				if (_unExpectedValue != null && _unExpectedValue.StartsWith(Constants.StateOverrides.StateToken))
+					setting = _unExpectedValue.Substring(Constants.StateOverrides.StateToken.Length);
+
+				string unExpectedValue = _unExpectedValue;
+				if (!String.IsNullOrEmpty(setting) &&
+					!savedState.TryGetValue(setting, out unExpectedValue))
+				{
+					throw new InvalidOperationException("OverrideUrl specified in definition but not found in state dictionary.  Did it depend on a request that failed?");
+				}
+
+				if (unExpectedValue != null && formattedActualValue.Equals(unExpectedValue))
```

Comment:
This is quite a lot of additional logic, and it could be simplified I think by splitting the logic. EndsWith, ExpectedValue, and UnexpectedValue are all mutually exclusive attributes. You could get rid of both EndsWith and UnexpectedValue, and instead have a separate attribute that describes how the value should be matched.

For example:

```
<StringProperty Name="Name" ExpectedValue="ValidatorTestFile.docx" IsRequired="true" MatchType="Equal" />
<StringProperty Name="Name" ExpectedValue=".docx" IsRequired="true" MatchType="EndsWith" />
<StringProperty Name="Name" ExpectedValue=".docx" IsRequired="true" MatchType="NotEqual" />
```

I think that would let you simplify the code above. Each 'mode' could be encompassed by a subclass of JsonStringPropertyValidator.